naga/src/front/wgsl/error.rs

```diff
@@ -763,26 +768,32 @@ impl<'a> Error<'a> {
                     .collect(),
                 notes: vec![],
             },
-            Error::InvalidSwitchValue { uint, span } => ParseError {
-                message: "invalid switch value".to_string(),
+            Error::InvalidSwitchSelector { span } => ParseError {
+                message: "invalid switch selector".to_string(),
                 labels: vec![(
                     span,
-                    if uint {
-                        "expected unsigned integer"
-                    } else {
-                        "expected signed integer"
-                    }
+                    "switch selector must be a scalar integer"
                     .into(),
                 )],
-                notes: vec![if uint {
-                    format!("suffix the integer with a `u`: `{}u`", &source[span])
-                } else {
-                    let span = span.to_range().unwrap();
-                    format!(
-                        "remove the `u` suffix: `{}`",
-                        &source[span.start..span.end - 1]
-                    )
-                }],
+                notes: vec![],
+            },
+            Error::InvalidSwitchCase { span } => ParseError {
+                message: "invalid switch case value".to_string(),
+                labels: vec![(
+                    span,
+                    "switch case selector must be a scalar integer const expression"
+                    .into(),
+                )],
+                notes: vec![],
+            },
+            Error::SwitchCaseTypeMismatch { span } => ParseError {
+                message: "invalid switch case selector value".to_string(),
+                labels: vec![(
+                    span,
+                    "switch case selector must have the same type as the switch selector expression"
+                    .into(),
+                )],
+                notes: vec![],
             },
```

Comment:
issue(non-blocking): As follow-up, we need to tell users _what_ the mismatched types are.

Reply:
Yes this would be good. Agreed it's best suited as a follow up, as it's not as easy as you might think, to do it well at least. You might want to show the span of the incorrect case, and show correct type, and additionally the span indicating where the correct type came from. But if the `n`th case statement has the incorrect type, the "correct" type might be implied by the `n-1`th case statement. It wouldn't be as simple as showing the span of the selector, as that might be abstract.

Reply:
Yep, that sounds similar to the domain of `match` statement arm type mismatches in Rust. Whee!

Reply:
Filed <https://github.com/gfx-rs/wgpu/issues/7293>.